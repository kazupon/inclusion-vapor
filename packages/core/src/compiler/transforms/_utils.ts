import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
