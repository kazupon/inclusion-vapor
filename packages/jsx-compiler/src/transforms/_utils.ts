import { parse } from '@babel/parser'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, JSXElement, JSXFragment } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const {
      body: [statement]
    } = parse(source, {
      sourceType: 'module',
      plugins: ['jsx']
    }).program
    let children!: JSXElement[] | JSXFragment['children']
    if (statement.type === 'ExpressionStatement') {
      children =
        statement.expression.type === 'JSXFragment'
          ? statement.expression.children
          : (statement.expression.type === 'JSXElement'
            ? [statement.expression]
            : [])
    }

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children,
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
