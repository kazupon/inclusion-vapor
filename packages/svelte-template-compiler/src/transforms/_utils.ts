import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, RootIRNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (
    source: string,
    overrideOptions: CompilerOptions = {}
  ): {
    ast: RootNode
    ir: RootIRNode
    code: string
    helpers: Set<string>
    vaporHelpers: Set<string>
  } => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
