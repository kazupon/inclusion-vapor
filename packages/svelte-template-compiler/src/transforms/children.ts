// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformChildren.ts

import { DynamicFlag, IRNodeTypes, isSvelteText } from '../ir/index.ts'
import { transformNode } from '../transform.ts'

import type { IRDynamicInfo, SvelteElement } from '../ir/index.ts'
import type { TransformContext } from './context.ts'
import type { NodeTransform } from './types.ts'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === IRNodeTypes.ROOT ||
    node.type === 'Fragment' ||
    node.type === 'InlineComponent' ||
    node.type === 'Slot'
  if (node.type !== 'Element' && !isFragment) {
    return
  }
  processChildren(node, context, isFragment)
}

export function processChildren(
  node: Parameters<NodeTransform>[0],
  context: Parameters<NodeTransform>[1],
  isFragment: boolean
): ReturnType<NodeTransform> {
  const children = node.children || []

  // normalize svelte text
  for (const [index, child] of [...children].entries()) {
    if (isSvelteText(child) && !child.data.trim()) {
      child.data = ' '
      if (!index) {
        children.splice(0, 1)
      } else if (index === children.length) {
        children.splice(-1, 1)
      }
    }
  }

  for (const [i, child] of children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childContext.dynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childContext.dynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    context.dynamic.children[i] = childContext.dynamic
  }

  if (!isFragment) {
    processDynamicChildren(context as TransformContext<SvelteElement>)
  }
}

function processDynamicChildren(context: TransformContext<SvelteElement>) {
  let prevDynamics: IRDynamicInfo[] = []
  let hasStaticTemplate = false
  const children = context.dynamic.children

  for (const [index, child] of children.entries()) {
    if (child.flags & DynamicFlag.INSERT) {
      prevDynamics.push(child)
    }

    if (!(child.flags & DynamicFlag.NON_TEMPLATE)) {
      if (prevDynamics.length > 0) {
        if (hasStaticTemplate) {
          context.childrenTemplate[index - prevDynamics.length] = `<!>`

          prevDynamics[0].flags -= DynamicFlag.NON_TEMPLATE
          const anchor = (prevDynamics[0].anchor = context.increaseId())

          context.registerOperation({
            type: IRNodeTypes.INSERT_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference(),
            anchor
          })
        } else {
          context.registerOperation({
            type: IRNodeTypes.PREPEND_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference()
          })
        }
        prevDynamics = []
      }
      hasStaticTemplate = true
    }
  }

  if (prevDynamics.length > 0) {
    context.registerOperation({
      type: IRNodeTypes.INSERT_NODE,
      elements: prevDynamics.map(child => child.id!),
      parent: context.reference()
    })
  }
}
