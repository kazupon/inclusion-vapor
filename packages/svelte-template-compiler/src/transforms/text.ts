// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformText.ts

import { createSimpleExpression } from '@vue-vapor/compiler-dom'
import {
  convertToSourceLocation,
  DynamicFlag,
  IRNodeTypes,
  isSvelteElement,
  isSvelteMustacheTag,
  isSvelteText
} from '../ir/index.ts'
import {
  getLiteralExpressionValue,
  isConstantExpression,
  resolveSimpleExpression
} from './utils.ts'

import type { SimpleExpressionNode } from '@vue-vapor/compiler-dom'
import type {
  RootNode,
  SvelteElement,
  SvelteMustacheTag,
  SvelteTemplateNode,
  SvelteText
} from '../ir/index.ts'
import type { TransformContext } from './context.ts'
import type { NodeTransform } from './types.ts'

type TextLike = SvelteText | SvelteMustacheTag

const seen = new WeakMap<TransformContext<RootNode>, WeakSet<SvelteTemplateNode | RootNode>>()

export const transformText: NodeTransform = (node, context) => {
  if (!seen.has(context.root)) {
    seen.set(context.root, new WeakSet())
  }

  if (seen.get(context.root)!.has(node)) {
    context.dynamic.flags |= DynamicFlag.NON_TEMPLATE
    return
  }

  const children = node.children || []
  if (isSvelteElement(node) && isAllTextLike(children)) {
    // TODO: should more tested for component
    processTextLikeContainer(children, context as TransformContext<SvelteElement>)
  } else if (isSvelteMustacheTag(node)) {
    // TODO: should more tested for MustacheTag expression
    // e.g. ConditionalExpression, LogicalExpression and CallExpression ...
    processTextLike(context as TransformContext<SvelteMustacheTag>)
  } else if (isSvelteText(node)) {
    context.template += node.data
  }
}

function processTextLike(context: TransformContext<SvelteMustacheTag>): void {
  const children = context.parent!.node.children || []
  const nexts = children.slice(context.index)
  const idx = nexts.findIndex(n => !isTextLike(n))
  const nodes = (idx > -1 ? nexts.slice(0, idx) : nexts) as Array<TextLike>
  const id = context.reference()
  const values = nodes.map(node => createTextLikeExpression(node, context))

  context.dynamic.flags |= DynamicFlag.INSERT | DynamicFlag.NON_TEMPLATE
  context.registerOperation({
    type: IRNodeTypes.CREATE_TEXT_NODE,
    id,
    values,
    // eslint-disable-next-line unicorn/no-array-callback-reference
    effect: !values.every(isConstantExpression) && !context.inVOnce
  })
}

function processTextLikeContainer(
  children: TextLike[],
  context: TransformContext<SvelteElement>
): void {
  const values = children.map(child => createTextLikeExpression(child, context))
  // eslint-disable-next-line unicorn/no-array-callback-reference
  const literals = values.map(getLiteralExpressionValue)
  if (literals.every(l => l != undefined)) {
    context.childrenTemplate = literals.map(String)
  } else {
    context.registerEffect(values, {
      type: IRNodeTypes.SET_TEXT,
      element: context.reference(),
      values
    })
  }
}

function createTextLikeExpression(node: TextLike, context: TransformContext): SimpleExpressionNode {
  seen.get(context.root)!.add(node)
  // prettier-ignore
  return isSvelteText(node)
    ? createSimpleExpression(node.data, true, convertToSourceLocation(node, (node as unknown as { raw: string }).raw || node.data))
    : resolveSimpleExpression(node, context as TransformContext<SvelteMustacheTag>)
}

function isAllTextLike(children: SvelteTemplateNode[]): children is TextLike[] {
  return (
    children.length > 0 &&
    // eslint-disable-next-line unicorn/no-array-callback-reference
    children.every(isTextLike) &&
    // at least one an mustache tag
    children.some(n => isSvelteMustacheTag(n))
  )
}

function isTextLike(node: SvelteTemplateNode): node is TextLike {
  // TODO: should more tested for MustacheTag expression
  // e.g. ConditionalExpression, LogicalExpression and CallExpression ...
  return isSvelteText(node) || isSvelteMustacheTag(node)
}
