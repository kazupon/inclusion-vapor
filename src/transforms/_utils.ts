import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type { RootIRNode as VaporRootIRNode } from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode } from '../ir'

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      prefixIdentifiers: true,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      prefixIdentifiers: true,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
