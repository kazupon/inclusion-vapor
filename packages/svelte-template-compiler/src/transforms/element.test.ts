import { BindingTypes, NodeTypes } from '@vue-vapor/compiler-dom'
import { compile as vaporCompile } from '@vue-vapor/compiler-vapor'
import { describe, expect, test } from 'vitest'
import { IRNodeTypes } from '../ir/index.ts'
import { makeCompile } from './_utils.ts'
import { transformChildren } from './children.ts'
import { transformElement } from './element.ts'
import { transformText } from './text.ts'

import type { BindingMetadata } from '@vue-vapor/compiler-dom'

const compileWithElementTransform = makeCompile({
  nodeTransforms: [transformText, transformElement, transformChildren]
})

describe('native elements', () => {
  test('basic', () => {
    const source = `
<div class="container">
  <div class="header">
    <p style="color: red;">Hello</p>
    <img src="foo.jpg" width="500" height="600">
    <form action="/submit" method="post">
      <input type="text" />
      <input type="submit">
    </form>
  </div>
</div>
`
    const { ir: _, code, vaporHelpers } = compileWithElementTransform(source)
    const expectedResult = vaporCompile(source)

    expect(code).toMatchSnapshot('svelte')
    expect(expectedResult.code).toMatchSnapshot('vue')

    // NOTE:
    // There are differences in the handling around spaces and line breaks between Vue compiler and Svelte compiler.
    // about details, see the snapshot
    // expect(code).toEqual(expectedResult.code)
    expect(vaporHelpers).toEqual(expectedResult.vaporHelpers)
  })

  test('static attributes', () => {
    const source = `<div id="foo" />`
    const { code, ir } = compileWithElementTransform(source)
    const expectedResult = vaporCompile(source)

    expect(code).toMatchSnapshot('svelte')
    expect(expectedResult.code).toMatchSnapshot('vue')

    expect(ir.template).toEqual([`<div id="foo"></div>`])
    expect(ir.block.effect).lengthOf(0)
  })

  test('attributes + children', () => {
    const source = `<div id="foo"><span/></div>`

    const { code, ir } = compileWithElementTransform(source)
    const expectedResult = vaporCompile(source)

    expect(code).toMatchSnapshot('svelte')
    expect(expectedResult.code).toMatchSnapshot('vue')

    expect(ir.template).toEqual([`<div id="foo"><span></span></div>`])
    expect(ir.block.effect).lengthOf(0)
  })

  test('static class', () => {
    const source = `<div class="foo">hello</div>`

    const { code, ir } = compileWithElementTransform(source)
    const expectedResult = vaporCompile(source)

    expect(code).toMatchSnapshot('svelte')
    expect(expectedResult.code).toMatchSnapshot('vue')

    expect(ir.template).toEqual([`<div class="foo">hello</div>`])
  })

  test('static style', () => {
    const source = `<div style="color: red;">hello</div>`

    const { code, ir } = compileWithElementTransform(source)
    const expectedResult = vaporCompile(source)

    expect(code).toMatchSnapshot('svelte')
    expect(expectedResult.code).toMatchSnapshot('vue')

    expect(ir.template).toEqual([`<div style="color: red;">hello</div>`])
  })
})

describe('component', () => {
  test('basic', () => {
    const source = `
<div class="container">
  <Foo class="foo"></Foo>
</div>
`
    const { ir: _, code, vaporHelpers } = compileWithElementTransform(source)
    const expectedResult = vaporCompile(source)

    expect(code).toMatchSnapshot('svelte')
    expect(expectedResult.code).toMatchSnapshot('vue')

    expect(code).toEqual(expectedResult.code)
    expect(vaporHelpers).toEqual(expectedResult.vaporHelpers)
  })

  test('import + resolve component', () => {
    const { code, ir, vaporHelpers } = compileWithElementTransform(`<Foo/>`)

    expect(code).toMatchSnapshot()
    expect(vaporHelpers).contains.all.keys('resolveComponent', 'createComponent')

    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        id: 0,
        tag: 'Foo',
        asset: true,
        root: true,
        props: [[]]
      }
    ])
  })

  test('resolve component from setup bindings', () => {
    const { code, ir, vaporHelpers } = compileWithElementTransform(`<Example />`, {
      bindingMetadata: {
        Example: BindingTypes.SETUP_MAYBE_REF
      }
    })

    expect(code).toMatchSnapshot()

    expect(vaporHelpers).not.toContain('resolveComponent')
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        tag: 'Example',
        asset: false
      }
    ])
  })

  test('resolve component from setup bindings (inline)', () => {
    const { code, vaporHelpers } = compileWithElementTransform(`<Example/>`, {
      inline: true,
      bindingMetadata: {
        Example: BindingTypes.SETUP_MAYBE_REF
      }
    })

    expect(code).toMatchSnapshot()

    expect(code).contains(`unref(Example)`)
    expect(vaporHelpers).not.toContain('resolveComponent')
    expect(vaporHelpers).toContain('unref')
  })

  test('resolve component from setup bindings (inline const)', () => {
    const { code, vaporHelpers } = compileWithElementTransform(`<Example/>`, {
      inline: true,
      bindingMetadata: {
        Example: BindingTypes.SETUP_CONST
      }
    })

    expect(code).toMatchSnapshot()

    expect(vaporHelpers).not.toContain('resolveComponent')
  })

  test('resolve namespaced component from setup bindings', () => {
    const { code, vaporHelpers } = compileWithElementTransform(`<Foo.Example/>`, {
      bindingMetadata: {
        Foo: BindingTypes.SETUP_MAYBE_REF
      }
    })

    expect(code).toMatchSnapshot()

    expect(code).contains(`_ctx.Foo.Example`)
    expect(vaporHelpers).not.toContain('resolveComponent')
  })

  test('resolve namespaced component from setup bindings (inline const)', () => {
    const { code, vaporHelpers } = compileWithElementTransform(`<Foo.Example/>`, {
      inline: true,
      bindingMetadata: {
        Foo: BindingTypes.SETUP_CONST
      }
    })

    expect(code).toMatchSnapshot()

    expect(code).contains(`Foo.Example`)
    expect(vaporHelpers).not.toContain('resolveComponent')
  })

  test('resolve namespaced component from props bindings (inline)', () => {
    const { code, vaporHelpers } = compileWithElementTransform(`<Foo.Example/>`, {
      inline: true,
      bindingMetadata: {
        Foo: BindingTypes.PROPS
      }
    })

    expect(code).toMatchSnapshot()

    expect(code).contains(`Foo.Example`)
    expect(vaporHelpers).not.toContain('resolveComponent')
  })

  test('resolve namespaced component from props bindings (non-inline)', () => {
    const { code, vaporHelpers } = compileWithElementTransform(`<Foo.Example/>`, {
      inline: false,
      bindingMetadata: {
        Foo: BindingTypes.PROPS
      }
    })

    expect(code).toMatchSnapshot()

    expect(code).contains('_ctx.Foo.Example')
    expect(vaporHelpers).not.toContain('resolveComponent')
  })

  test('do not resolve component from non-script-setup bindings', () => {
    const bindingMetadata: BindingMetadata = {
      Example: BindingTypes.SETUP_MAYBE_REF
    }
    Object.defineProperty(bindingMetadata, '__isScriptSetup', {
      value: false
    })
    const { code, ir, vaporHelpers } = compileWithElementTransform(`<Example/>`, {
      bindingMetadata
    })

    expect(code).toMatchSnapshot()

    expect(vaporHelpers).toContain('resolveComponent')
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        id: 0,
        tag: 'Example',
        asset: true
      }
    ])
  })

  test('generate single root component', () => {
    const { code } = compileWithElementTransform(`<Comp/>`, {
      bindingMetadata: { Comp: BindingTypes.SETUP_CONST }
    })

    expect(code).toMatchSnapshot()

    expect(code).contains('_createComponent(_ctx.Comp, null, null, true)')
  })

  test('generate multi root component', () => {
    const { code } = compileWithElementTransform(`<Comp/>123`, {
      bindingMetadata: { Comp: BindingTypes.SETUP_CONST }
    })

    expect(code).toMatchSnapshot()

    expect(code).contains('_createComponent(_ctx.Comp)')
  })

  describe('dynamic component', () => {
    test('dynamic binding', () => {
      const { code, ir, vaporHelpers } = compileWithElementTransform(
        `<svelte:component this={foo}></svelte:component>`
      )
      expect(code).toMatchSnapshot()
      expect(vaporHelpers).toContain('resolveDynamicComponent')
      expect(ir.block.operation).toMatchObject([
        {
          type: IRNodeTypes.CREATE_COMPONENT_NODE,
          tag: 'component',
          asset: true,
          root: true,
          // TODO:
          // props: [[]],
          dynamic: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: 'foo',
            isStatic: false
          }
        }
      ])
    })
  })

  test('static props', () => {
    const source = `
<Foo id="foo" class="bar" />
`

    const { code, ir } = compileWithElementTransform(source)
    const expectedResult = vaporCompile(source)

    expect(code).toMatchSnapshot('svelte')
    expect(expectedResult.code).toMatchSnapshot('vue')

    expect(code).toEqual(expectedResult.code)

    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        tag: 'Foo',
        asset: true,
        root: true,
        props: [
          [
            {
              key: {
                type: NodeTypes.SIMPLE_EXPRESSION,
                content: 'id',
                isStatic: true
              },
              values: [
                {
                  type: NodeTypes.SIMPLE_EXPRESSION,
                  content: 'foo',
                  isStatic: true
                }
              ]
            },
            {
              key: {
                type: NodeTypes.SIMPLE_EXPRESSION,
                content: 'class',
                isStatic: true
              },
              values: [
                {
                  type: NodeTypes.SIMPLE_EXPRESSION,
                  content: 'bar',
                  isStatic: true
                }
              ]
            }
          ]
        ]
      }
    ])
  })
})

test('empty template', () => {
  const { code } = compileWithElementTransform('')

  expect(code).toMatchSnapshot()
  expect(code).contain('return null')
})

test('scoped css', () => {
  const source = `
<div class="container">
  <div class="header">
    <p style="color: red;">Hello</p>
    <img src="foo.jpg" width="500" height="600">
    <form action="/submit" method="post">
      <input type="text" />
      <input type="submit">
    </form>
  </div>
</div>
<style>
.container {
  color: blue;
}
p {
  color: green;
}
div.header {
  color: yellow;
}
:global(from) {
  color: purple;
}
</style>
`
  const {
    ir: _,
    code,
    vaporHelpers: __
  } = compileWithElementTransform(source, { scopeId: 'svelte-xxx' })

  expect(code).toMatchSnapshot()
})
