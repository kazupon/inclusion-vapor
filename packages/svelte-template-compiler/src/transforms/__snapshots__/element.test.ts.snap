// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`component > basic > svelte 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent, insert as _insert, template as _template } from 'vue/vapor';
const t0 = _template("<div class=\\"container\\"></div>")

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const n1 = t0()
  const n0 = _createComponent(_component_Foo, [
    { class: () => ("foo") }
  ])
  _insert(n0, n1)
  return n1
}"
`;

exports[`component > basic > vue 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent, insert as _insert, template as _template } from 'vue/vapor';
const t0 = _template("<div class=\\"container\\"></div>")

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const n1 = t0()
  const n0 = _createComponent(_component_Foo, [
    { class: () => ("foo") }
  ])
  _insert(n0, n1)
  return n1
}"
`;

exports[`component > do not resolve component from non-script-setup bindings 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Example = _resolveComponent("Example")
  const n0 = _createComponent(_component_Example, null, null, true)
  return n0
}"
`;

exports[`component > generate multi root component 1`] = `
"import { createComponent as _createComponent, template as _template } from 'vue/vapor';
const t0 = _template("123")

export function render(_ctx) {
  const n1 = t0()
  const n0 = _createComponent(_ctx.Comp)
  return [n0, n1]
}"
`;

exports[`component > generate single root component 1`] = `
"import { createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createComponent(_ctx.Comp, null, null, true)
  return n0
}"
`;

exports[`component > import + resolve component 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const n0 = _createComponent(_component_Foo, null, null, true)
  return n0
}"
`;

exports[`component > resolve component from setup bindings (inline const) 1`] = `
"(() => {
  const n0 = _createComponent(Example, null, null, true)
  return n0
})()"
`;

exports[`component > resolve component from setup bindings (inline) 1`] = `
"(() => {
  const n0 = _createComponent(_unref(Example), null, null, true)
  return n0
})()"
`;

exports[`component > resolve component from setup bindings 1`] = `
"import { createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createComponent(_ctx.Example, null, null, true)
  return n0
}"
`;

exports[`component > resolve namespaced component from props bindings (inline) 1`] = `
"(() => {
  const n0 = _createComponent(Foo.Example, null, null, true)
  return n0
})()"
`;

exports[`component > resolve namespaced component from props bindings (non-inline) 1`] = `
"import { createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createComponent(_ctx.Foo.Example, null, null, true)
  return n0
}"
`;

exports[`component > resolve namespaced component from setup bindings (inline const) 1`] = `
"(() => {
  const n0 = _createComponent(Foo.Example, null, null, true)
  return n0
})()"
`;

exports[`component > resolve namespaced component from setup bindings 1`] = `
"import { createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const n0 = _createComponent(_ctx.Foo.Example, null, null, true)
  return n0
}"
`;

exports[`component > static props > svelte 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const n0 = _createComponent(_component_Foo, [
    {
      id: () => ("foo"), 
      class: () => ("bar")
    }
  ], null, true)
  return n0
}"
`;

exports[`component > static props > vue 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent } from 'vue/vapor';

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const n0 = _createComponent(_component_Foo, [
    {
      id: () => ("foo"), 
      class: () => ("bar")
    }
  ], null, true)
  return n0
}"
`;

exports[`empty template 1`] = `
"
export function render(_ctx) {
  return null
}"
`;

exports[`native elements > attributes + children > svelte 1`] = `
"import { template as _template } from 'vue/vapor';
const t0 = _template("<div id=\\"foo\\"><span></span></div>")

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;

exports[`native elements > attributes + children > vue 1`] = `
"import { template as _template } from 'vue/vapor';
const t0 = _template("<div id=\\"foo\\"><span></span></div>")

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;

exports[`native elements > basic > svelte 1`] = `
"import { template as _template } from 'vue/vapor';
const t0 = _template("<div class=\\"container\\"><div class=\\"header\\"><p style=\\"color: red;\\">Hello</p> <img src=\\"foo.jpg\\" width=\\"500\\" height=\\"600\\"> <form action=\\"/submit\\" method=\\"post\\"><input type=\\"text\\"> <input type=\\"submit\\"></form></div></div>")

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;

exports[`native elements > basic > vue 1`] = `
"import { template as _template } from 'vue/vapor';
const t0 = _template("<div class=\\"container\\"><div class=\\"header\\"><p style=\\"color: red;\\">Hello</p><img src=\\"foo.jpg\\" width=\\"500\\" height=\\"600\\"><form action=\\"/submit\\" method=\\"post\\"><input type=\\"text\\"><input type=\\"submit\\"></form></div></div>")

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;

exports[`native elements > static attributes > svelte 1`] = `
"import { template as _template } from 'vue/vapor';
const t0 = _template("<div id=\\"foo\\"></div>")

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;

exports[`native elements > static attributes > vue 1`] = `
"import { template as _template } from 'vue/vapor';
const t0 = _template("<div id=\\"foo\\"></div>")

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;

exports[`native elements > static class > svelte 1`] = `
"import { template as _template } from 'vue/vapor';
const t0 = _template("<div class=\\"foo\\">hello</div>")

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;

exports[`native elements > static class > vue 1`] = `
"import { template as _template } from 'vue/vapor';
const t0 = _template("<div class=\\"foo\\">hello</div>")

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;
